### PR TITLE
plugins/yazi: init

### PR DESCRIPTION
**Pull request by khaneliman**
Adding module for https://github.com/mikavilpas/yazi.nvim

NOTE: It looks like it has a dependency on telescope in the builtin functions for searching. Should that be a warning or added as an extra plugin automatically? To make this more complex, technically a person can setup the integrations property to change the grep behavior to not rely on it..

### CODE REVIEW

**Comment by GaetanLepage** [comment]
The developer doesn't list telescope as a dependency. In that sense, I don't think that we should do anything.
Can the plugin be operated normally when simply enabling `plugins.yazi.enable` ?

**Comment by khaneliman** [comment]
> The developer doesn't list telescope as a dependency. In that sense, I don't think that we should do anything. Can the plugin be operated normally when simply enabling `plugins.yazi.enable` ?

Yeah, it's fine to use most of the functionality. [Here's](https://github.com/mikavilpas/yazi.nvim/blob/4e549120e1ff2702b6ea1eff3640a879fffb2141/lua/yazi/config.lua#L46-L66) where I'm talking about.

**Comment by khaneliman** [comment]
Added a description to make note about it.

**Comment by MattSturgeon** [comment]
> The warning would basically have to be checking if the user hasn't overridden the `settings.integrations.grep_in_directory` then check if telescope is enabled or something and even that doesn't feel foolproof.

TBH this seems like something that upstream plugin should be dealing with. If the user configures the plugin with a mapping that needs telescope (even if it's a default mapping) the plugin should really warn that telescope is missing.

Perhaps we should report an issue upstream?

File: tests/test-sources/plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline tests/test-sources/plugins/filetrees/yazi.nix:64]:
```suggestion
```

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:9]:
```suggestion
```

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:7]: Do we need `with lib` here? I know we've done this historically, but I'd rather inherit the few things we use repeatedly and access the rest explicitly.

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:27]:
```suggestion
    You can override the grep functionality through the `plugins.yazi.settings.integrations.grep_in_directory`
    and `plugins.yazi.settings.integrations.grep_in_selected_files` options
    if you do not wish to use `telescope`.
    Otherwise you must enable `plugins.telescope` or install it some other way.
```

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:142]: Nixfmt has done a _horrible_ job here!

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:160]: @GaetanLepage is this a dependency that should be upstreamed to nixpkgs? Or perhaps it's already handled there?

File: plugins/filetrees/yazi.nix
Comment (khaneliman) [reply plugins/filetrees/yazi.nix:141]: Yeah... I didn't know if this was another situation where we needed to use `'` and set the options directly.

Comment (khaneliman) [reply plugins/filetrees/yazi.nix:141]:
```nix
    yazi_floating_window_winblend = defaultNullOpts.mkNullableWithRaw' {
      type = (types.ints.between 0 100);
      pluginDefault = 0;
      description = "`0` for fully opaque and `100` for fully transparent. See :h winblend";
    };
```

File: plugins/filetrees/yazi.nix
Comment (khaneliman) [reply plugins/filetrees/yazi.nix:159]: I didn't add it when i added the plugin. Is that just adding a dependency in ovverrides.nix?
```nix
  yazi-nvim = super.yazi-nvim.overrideAttrs {
    dependencies = with self; [ plenary-nvim ];
  };
```

Comment (MattSturgeon) [reply plugins/filetrees/yazi.nix:159]: > Is that just adding a dependency in overrides.nix?

Yes, I think so. Though @GaetanLepage is more familiar than I am with nixpkgs's vim/nvim contributions.

Comment (khaneliman) [reply plugins/filetrees/yazi.nix:162]: Created a PR for the dependency upstream and added a TODO.

**Comment by khaneliman** [comment]
> > The warning would basically have to be checking if the user hasn't overridden the `settings.integrations.grep_in_directory` then check if telescope is enabled or something and even that doesn't feel foolproof.
> 
> TBH this seems like something that upstream plugin should be dealing with. If the user configures the plugin with a mapping that needs telescope (even if it's a default mapping) the plugin should really warn that telescope is missing.
> 
> Perhaps we should report an issue upstream?

Looking at the documentation on their README, it is noted in their that it uses telescope when it's available otherwise the binding does nothing.

**Comment by MattSturgeon** [comment]
> Looking at the documentation on their README, it is noted in their that it uses telescope when it's available otherwise the binding does nothing. [mikavilpas/yazi.nvim#%EF%B8%8F-keybindings](https://github.com/mikavilpas/yazi.nvim?tab=readme-ov-file#%EF%B8%8F-keybindings). I think we should be fine with the description text, link to their documentation, and their documentation calling it out, too.

Well spotted. Following that link, it seems `<c-g>` has a similar dependency on [grug-far](https://github.com/MagicDuck/grug-far.nvim).

If both of these fail somewhat gracefully, as the documentation would imply, and on the basis that you've been using the plugin a while without even noticing the optional deps, I think we can reword or remove our note in the description?

**Comment by khaneliman** [comment]
> > Looking at the documentation on their README, it is noted in their that it uses telescope when it's available otherwise the binding does nothing. [mikavilpas/yazi.nvim#%EF%B8%8F-keybindings](https://github.com/mikavilpas/yazi.nvim?tab=readme-ov-file#%EF%B8%8F-keybindings). I think we should be fine with the description text, link to their documentation, and their documentation calling it out, too.
> 
> Well spotted. Following that link, it seems `<c-g>` has a similar dependency on [grug-far](https://github.com/MagicDuck/grug-far.nvim).
> 
> If both of these fail somewhat gracefully, as the documentation would imply, and on the basis that you've been using the plugin a while without even noticing the optional deps, I think we can reword or remove our note in the description?

Updated description to be more clear about the external dependencies. Let me know if it's clear enough.

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:33]: Would it be easier long term to just link to the upstream docs?

Something like

```md
Some default keybindings have additional dependencies you may need to install or enable.
See the [upstream docs](https://github.com/mikavilpas/yazi.nvim?tab=readme-ov-file#%EF%B8%8F-keybindings) for details.
```

Comment (khaneliman) [reply plugins/filetrees/yazi.nix:33]: Yeah, fair enough. Less to maintain and keeps the ownership on upstream.

**Review by MattSturgeon — APPROVED** [review]
LGTM. Thanks again!

File: plugins/filetrees/yazi.nix
Comment (MattSturgeon) [inline plugins/filetrees/yazi.nix:159]: I have a feeling #333506 will merge soon as it's already got one approval. Shall we just wait until we update to a nixpkgs rev with the change before merging so we don't have to drop this line in another PR?

Don't really mind either way.

Comment (khaneliman) [reply plugins/filetrees/yazi.nix:159]: Usually a couple days delay before getting to unstable after merge and then whenever we get to flake update. I dont mind keeping track of the nixpkgs PR and when we do another flake.lock update. I usually do these kind of TODO's in my own config with overlays while waiting for fixes from nixos-unstable.

**Comment by MattSturgeon** [comment]
@mergifyio queue

**Pull request merged**
